### PR TITLE
Add weight decay, linear scheduler, and brackets to [../saved.models]

### PR DESCRIPTION
**Pull request by amunozo**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -81,6 +84,26 @@ def save_folder_name(self, given_run_name):
...
+        seed = random.randint(0, 2**32 - 1)
+        print(f"No seed provided. Using generated seed: {seed}")
+    else:
+        print(f"Using provided seed: {seed}")
+    
+    pl.seed_everything(seed)
+    
+    torch.manual_seed(seed)
+    torch.cuda.manual_seed(seed)
+    torch.cuda.manual_seed_all(seed)
```
Comment (gailweiss) [inline main.py:98]: can you set the torch mps random seed as well please?

File: train/train_params.py [hunk 2]
```diff
@@ -79,6 +79,10 @@ class TrainParams:
 #           learning rate lr and reduces the learning rate by scheduler_factor
 #           every time the train batch loss ceases improving for patience
 #           steps, reducing the learning rate only up to min_lr.
+#       "Linear":
+#           Uses torch.optim.lr_scheduler.LinearLR. Starts from lr and decreases
+#           the learning rate linearly down to min_lr over the course of
+#           training steps, as defined by the total training duration.
```
[the remark this answers is not in the document]
Comment (gailweiss) [reply train/train_params.py:87]: still missing a clarification of what "training duration" is :)

File: train/trainer.py [hunk 3]
```diff
@@ -219,25 +219,31 @@ def make_main_scheduler(self, optimizer):
...
         elif self.train_params.lr_scheduler_type == 'Cosine':
             sched = torch.optim.lr_scheduler.CosineAnnealingLR
             return sched(optimizer, self.train_params.lr_cycle_steps,
-                         eta_min=self.train_params.min_lr)
+                        eta_min=self.train_params.min_lr)
+        elif self.train_params.lr_scheduler_type == 'Linear':
+            sched = torch.optim.lr_scheduler.LinearLR
+            return sched(optimizer, start_factor=1.0,
+                        end_factor=self.train_params.min_lr / self.train_params.lr,
+                        total_iters=self.train_params.epochs * self.train_params.accumulate_grad_batches)
```
[the remark this answers is not in the document]
Comment (gailweiss) [reply train/trainer.py:242]: still unclear to me here how this will operate. note that my code steps the scheduler at every training step! still not clear what accumulate_grad_batches is doing in this code

File: main.py [hunk 4]
```diff
@@ -237,6 +236,7 @@ def save_model(args, saving_folder, pltrainer, dp, tp):
 
 
 def run_config(args, dp, tp, mp, namer):
+    seed = seed_everything(args.seed, tp)
```
Comment (gailweiss) [inline main.py:239]: why does seed_everything return the seed, and why are you getting it?

File: train/train_params.py [hunk 5]
```diff
@@ -105,4 +106,6 @@ class TrainParams:
 #   Relevant only when also passing --save through the args, otherwise will be
 #   overridden to 0.
 # early_stop_nsamples:
-#   When >0, the number of batches to train before early stopping the training.
\ No newline at end of file
+#   When >0, the number of batches to train before early stopping the training.
+# seed:
+#   Random seed for reproducibility
```
Comment (gailweiss) [inline train/train_params.py:111]: write that it can be overwritten by main's args. also write that it is set randomly if None, and then stored with the randomly set value. state that the default value is None. do a run and verify that indeed this happens (i.e. that it sets a random value if None, that it stores it properly, and that you can then go get this value from the logs and use it to make an identical training run)

File: train/train_params.py [hunk 6]
```diff
@@ -25,6 +25,7 @@ class TrainParams:
     checkpoint_every: int = 0
     early_stop_nsamples: int = -1
     weight_decay: float = 0.01
+    seed: int
```
Comment (gailweiss) [inline train/train_params.py:28]: give default value (None), i.e.  seed: int = None

File: train/train_params.py [hunk 7]
```diff
@@ -24,6 +24,7 @@ class TrainParams:
     lora_std: float = 0.02
     checkpoint_every: int = 0
     early_stop_nsamples: int = -1
+    weight_decay: float = 0.01
```
[the remark this answers is not in the document]
Comment (gailweiss) [reply train/train_params.py:27]: still need description of weight_decay in file comments
